lib/map_walker.go

```diff
@@ -0,0 +1,162 @@
+package lib
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/mitchellh/copystructure"
+	"github.com/mitchellh/mapstructure"
+	"github.com/mitchellh/reflectwalk"
+)
+
+func MapWalk(input interface{}) (map[string]interface{}, error) {
+	mapCopyRaw, err := copystructure.Copy(input)
+	if err != nil {
+		return nil, err
+	}
+
+	mapCopy, ok := mapCopyRaw.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("internal error: type used in MapWalk is not a map[string]interface{}")
+	}
+
+	if err := reflectwalk.Walk(mapCopy, &mapWalker{}); err != nil {
+		return nil, err
+	}
+
+	return mapCopy, nil
+}
+
+var typMapIfaceIface = reflect.TypeOf(map[interface{}]interface{}{})
+
+// mapWalker implements interfaces for the reflectwalk package
+// (github.com/mitchellh/reflectwalk) that can be used to automatically
+// make a JSON compatible map safe for JSON usage. This is currently
+// targeted at the map[string]interface{}
+//
+// Most of the implementation here is just keeping track of where we are
+// in the reflectwalk process, so that we can replace values. The key logic
+// is in Slice() and SliceElem().
+//
+// In particular we're looking to replace two cases the msgpack codec causes:
+//
+//   1.) String values get turned into byte slices. JSON will base64-encode
+//       this and we don't want that, so we convert them back to strings.
```

Comment:
I've never quite understood why we have to manually fix this - go-msgpack handles have a `RawToString` option that in most places we enable that causes strings to be interpreted as string rather than `[]byte`. Is this setting ignored when decoding into an `interface{}`? It seems like its the whole point of that flag to fix this issue but maybe there is more subtlety there I'm not seeing.

Reply:
RawToString would probably fix the []byte -> string issues. However with arbitrarily nested JSON compatible config there is the second case of where msgpack will decode into a map[interface{}]interface{} (supposedly according to all the places we have done this in the past)

Reply:
That doesn't apply here since this is getting used for json decoding around the api, right? The equivalent would be a custom mapstructure decode hook (or the weakly typed conversions). I think we could use that instead of needing to do the mapwalk here to convert any []uint8s. We do a similar thing here with a custom decode hook in the CA config: https://github.com/hashicorp/consul/blob/ef9f27cbc8f591704f9983afa3efac07a2b07902/agent/structs/connect_ca.go#L354-L386

Reply:
A custom decode hook could potentially solve the problem but many of the places where this gets decoded we are not using mapstructure such as when calling `structs.Decode`. There we just do a msgpack decode.